openssl-dynamic/src/main/c/sslutils.c

```diff
@@ -329,37 +327,51 @@ static unsigned char dhxxx2_g[]={
 static DH *get_dh(int idx)
 {
     DH *dh;
-    if ((dh = DH_new()) == NULL)
+    BIGNUM *p, *g;
+    if ((dh = DH_new()) == NULL) {
         return NULL;
+    }
+
     switch (idx) {
         case SSL_TMP_KEY_DH_512:
-            dh->p = BN_bin2bn(dh0512_p, sizeof(dh0512_p), NULL);
+            p = BN_bin2bn(dh0512_p, sizeof(dh0512_p), NULL);
         break;
         case SSL_TMP_KEY_DH_1024:
-            dh->p = BN_bin2bn(dh1024_p, sizeof(dh1024_p), NULL);
+            p = BN_bin2bn(dh1024_p, sizeof(dh1024_p), NULL);
         break;
         case SSL_TMP_KEY_DH_2048:
-            dh->p = BN_bin2bn(dh2048_p, sizeof(dh2048_p), NULL);
+            p = BN_bin2bn(dh2048_p, sizeof(dh2048_p), NULL);
         break;
         case SSL_TMP_KEY_DH_4096:
-            dh->p = BN_bin2bn(dh4096_p, sizeof(dh2048_p), NULL);
+            p = BN_bin2bn(dh4096_p, sizeof(dh2048_p), NULL);
         break;
     }
-    dh->g = BN_bin2bn(dhxxx2_g, sizeof(dhxxx2_g), NULL);
-    if ((dh->p == NULL) || (dh->g == NULL)) {
-        DH_free(dh);
-        return NULL;
+    if (p == NULL) {
+        goto error;
     }
-    else
-        return dh;
-    return NULL;
-}
+
+    g = BN_bin2bn(dhxxx2_g, sizeof(dhxxx2_g), NULL);
+    if (g == NULL) {
+        goto error;
+    } else {
+// DH_set0_pqg() was introduced to initialize the DH parameters in OpenSSL 1.1.0 and DH is opaque now.
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
+        if (DH_set0_pqg(dh, p, NULL, g) == 0) {
+            goto error;
+        }
 #else
-static DH *get_dh(int idx)
-{
+        dh->p = p;
+        dh->g = g;
+#endif
+        return dh;
+    }
+error:
+    // All these functions can handle NULL as argument.
```

Comment:
But couldn't p and g be uninitialized?

Reply:
@ejona86 sorry I don't get the question... All the free methods can handle `NULL` so it should be fine, no ?

Reply:
Above:
```
BIGNUM *p, *g;
```

I'd expect `p` and `g` need to be set to `NULL` to initialize them.

Reply:
ah yeah better be safe here... Thanks!